src/dbus.c

```diff
@@ -17,106 +17,48 @@
-    "        because this allows to implement a Qt version if desired without frictions."
-    "        -->"
-    "    <interface name=\"io.snapcraft.SnapDesktopIntegration\">"
-    "        <!--"
-    "            ApplicationIsBeingRefreshed:"
-    "            @application_name: the name of the application name"
-    "            @lock_file: the full path to the lock file, to allow"
-    "                        the daemon to detect when the window must"
-    "                        be closed, or an empty string to not monitor"
-    "                        any lock file."
-    "           @extra_parameters: a dictionary with extra optional parameters."
-    "                              Currently defined parameters are:"
-    "                * icon: a string with the path to an icon."
-    "           Method used to notify to the user that the snap that"
-    "           they wanted to run is being refreshed, and they have"
-    "           to wait until the process has finished."
-    "       -->"
-    "       <method name=\"ApplicationIsBeingRefreshed\">"
-    "       <arg direction=\"in\" type=\"s\" name=\"application_name\"/>"
-    "       <arg direction=\"in\" type=\"s\" name=\"lock_file\"/>"
-    "       <arg direction=\"in\" type=\"a{sv}\" name=\"extra_parameters\"/>"
-    "        </method>"
-    "       <!--"
-    "            ApplicationRefreshCompleted:"
-    "            @application_name: the name of the application name"
-    "           @extra_parameters: a dictionary with extra optional parameters."
-    "           Method used to notify the daemon that the refresh is completed"
-    "           and the dialog can be closed."
-    "       -->"
-    "       <method name=\"ApplicationRefreshCompleted\">"
-    "       <arg direction=\"in\" type=\"s\" name=\"application_name\"/>"
-    "       <arg direction=\"in\" type=\"a{sv}\" name=\"extra_parameters\"/>"
-    "        </method>"
-    "    </interface>"
-    "</node>";
-
-    g_autoptr(GDBusNodeInfo) node_info = g_dbus_node_info_new_for_xml (introspection_xml, error);
-    if (node_info == NULL) {
-        g_prefix_error(error, "Failed to parse D-Bus XML: ");
-        return FALSE;
-    }
 
-    static const GDBusInterfaceVTable notifications_vtable = {
-        handle_notifications_method_call,
-        NULL,
-        NULL,
-    };
-    if (g_dbus_connection_register_object(connection,
-                                          "/io/snapcraft/SnapDesktopIntegration",
-                                          node_info->interfaces[0],
-                                          &notifications_vtable,
-                                          (gpointer) state,
-                                          NULL,
-                                          error) == 0) {
-        g_prefix_error(error, "Failed to register notifications object: ");
-        return FALSE;
-   }
-   return TRUE;
+    SnapDesktopIntegration *skeleton = snap_desktop_integration_skeleton_new();
```

Comment:
It looks like this object is leaked. Perhaps return the object, store it in DsState, and unref it in ds_state_free()?

Reply:
Not really. This object has to live from the beginning to the end of the program, because it is the one that manages the DBus connections. DsState, instead, is the state of each progress window: one is created every time a progress window is shown, and destroyed when it is closed.

Of course, I can store the skeleton somewhere to unref() it when the program ends...

Reply:
There's just one `DsState` created in `main` and freed when main exits. That seems like the appropriate lifetime for the skeleton. Are you thinking of RefreshState here?

Reply:
Oh... you are right...